iceoryx_hoofs/include/iceoryx_hoofs/internal/design_pattern/polymorphic_handler.inl

```diff
@@ -0,0 +1,163 @@
+{
+    m_active = true;
+}
+
+void Activatable::deactivate() noexcept
+{
+    m_active = false;
+}
+
+bool Activatable::isActive() const noexcept
+{
+    return m_active;
+}
+
+// The get method is considered to be on the hot path and works as follows:
+//
+// on first call (in a thread):
+// 1. localHandler is initialized
+//    - getCurrent is called
+//    - instantiates singleton instance()
+//    - instantiates default
+//    - sets m_current of instance to default instance (release)
+//    - default is active
+
+// if any thread changes the active handler with set (or reset)
+// under mutex protection, it will:
+//    - set the new handler to active
+//    - set current handler to the new handler
+//    - deactivate the old handler (can still be used as it still needs to exist)
+//
+
+// on any call after the handler was changes in another thread
+// 1. we check whether the handler is active
+// (can be outdated information but will eventually be false once the atomic value is updated)
+// 2. if it was changed it is now inactive and we wait to obtain the mutex lock
+// under lock, we update the local handler to the new one (note that it cannot change
+// while this happens as we hold the lock)
+
+template <typename Interface, typename Default, typename Hooks>
+Interface& PolymorphicHandler<Interface, Default, Hooks>::get() noexcept
+{
+    // NOLINTNEXTLINE(cppcoreguidelines-avoid-non-const-global-variables) false positive, thread_local
+    thread_local Interface* localHandler = getCurrent(); // initialized once per thread on first call
+
+    if (!localHandler->isActive())
+    {
+        std::lock_guard<std::mutex> lock(instance().m_mutex);
+        localHandler = getCurrent();
+    }
+    return *localHandler;
+}
+
+// can throw
+template <typename Interface, typename Default, typename Hooks>
+Interface* PolymorphicHandler<Interface, Default, Hooks>::set(Interface& handler) noexcept
+{
+    auto& ins = instance();
+
+    // setting is rare and we lock to synchronize and update the active flags properly
+    std::lock_guard<std::mutex> lock(ins.m_mutex);
+    if (ins.m_isFinal)
+    {
+        Hooks::onSetAfterFinalize(*getCurrent(), handler);
+        return nullptr;
+    }
+
+    handler.activate(); // it may have been deactivated before, so always reactivate it
+    auto prev = ins.m_current.exchange(&handler, std::memory_order_relaxed);
+
+    // anyone still using it will eventually see that it is inactive
+    // and switch to the new handler
+    prev->deactivate();
+    return prev;
+}
+
+template <typename Interface, typename Default, typename Hooks>
+Interface* PolymorphicHandler<Interface, Default, Hooks>::reset() noexcept
+{
+    return set(getDefault());
+}
+
+template <typename Interface, typename Default, typename Hooks>
+void PolymorphicHandler<Interface, Default, Hooks>::finalize() noexcept
+{
+    auto& ins = instance();
+    std::lock_guard<std::mutex> lock(ins.m_mutex);
+    ins.m_isFinal.store(true);
+}
```

Comment:
The question is if this would become obsolete. Probably not since one might finalize without setting a new object.

This might lead to a race though if not done carefully. I have an idea how to do it but don't want to take you the fun of thinking about it :)

Reply:
I would keep it to freeze the current handler, even to the default one.